src/iterative/KOSMAPOSL/KOSMAPOSLReconstruction.cxx

```diff
@@ -0,0 +1,1487 @@
+//                                        - (*anatomical_prior_sptr)[z+dz][y+dy][x+dx])
+//                                       / anatomical_sd
+//                                       / sigma_m)
+//                               / 2)
+//                         * exp(-square(distance[dz][dy][dx] / grid_spacing.x())
+//                               / (2 * square(sigma_dm)));
+
+//                       // double anatomical_kernel = calc_anatomical_kernel(
+//                       //   x, y, z, min_x, min_y, min_z, max_x, max_y, max_z,
+//                       //   dx, dy, dz, min_dx, min_dy, min_dz, max_dx, max_dy, max_dz,
+//                       //   true, grid_spacing, distance);
+//                       // if (kanatomical - anatomical_kernel > 1e-3) {
+//                       //   std::cout << "x: " << x << ", y: " << y << ", z: " << z << std::endl;
+//                       //   std::cout << "dx: " << dx << ", dy: " << dy << ", dz: " << dz << std::endl;
+//                       //   std::cout << "kanatomical: " << kanatomical;
+//                       //   std::cout << ", anatomical_kernel; " << anatomical_kernel << std::endl;
+//                       // }
+
+//                       pnkernel += kPET*kanatomical;
+
+//                       kernelised_image_out[z][y][x]
+//                         += kanatomical*kPET*image_to_kernelise[z+dz][y+dy][x+dx];
+//                     }
+
+//               if (current_alpha_estimate[z][y][x]==0) {
+//                 continue;
+//               }
+
+//               kernelised_image_out[z][y][x]
+//                 = kernelised_image_out[z][y][x] / pnkernel;
+//             }
+//         }
+//     }
+
+// }
+
+template<typename TargetT>
+void KOSMAPOSLReconstruction<TargetT>::compute_kernelised_image(
+                         TargetT& kernelised_image_out,
+                         const TargetT& image_to_kernelise,
+                         const TargetT& current_alpha_estimate)
+{
+  // if (this->num_non_zero_feat==1) {
+  //   compact_compute_kernelised_image(kernelised_image_out, image_to_kernelise, current_alpha_estimate);
+  // }
+  // else {
+  //   full_compute_kernelised_image(kernelised_image_out, image_to_kernelise, current_alpha_estimate);
+  // }
+  // return;
+
+  bool use_compact_implementation = this->num_non_zero_feat == 1;
+
+  // Something very weird happens here if I do not get_empty_copy()
+  // KImage elements will be all nan
+  unique_ptr<TargetT> kImage_uptr(current_alpha_estimate.get_empty_copy());
+  kernelised_image_out = *kImage_uptr;
+
+  const DiscretisedDensityOnCartesianGrid<3,float>* current_anatomical_cast =
+    dynamic_cast< const DiscretisedDensityOnCartesianGrid<3,float> *>
+      (this->get_anatomical_prior_sptr ().get());
+  // TODO - which spacing to use? Need both?
+  const CartesianCoordinate3D<float>& grid_spacing =
+    current_anatomical_cast->get_grid_spacing();
+
+  // std::cout << "a" << std::endl;
+  Array<3, float> distance
```

Comment:
constant?

Reply:
why constant? this should change for every voxel